src/java/org/apache/cassandra/streaming/async/StreamingInboundHandler.java

```diff
@@ -232,24 +225,9 @@ else if (t instanceof StreamReceiveException)
 
         StreamSession deriveSession(StreamMessage message)
         {
-            StreamSession streamSession = null;
             // StreamInitMessage starts a new channel, and IncomingStreamMessage potentially, as well.
             // IncomingStreamMessage needs a session to be established a priori, though
-            if (message instanceof StreamInitMessage)
-            {
-                assert session == null : "initiator of stream session received a StreamInitMessage";
-                StreamInitMessage init = (StreamInitMessage) message;
-                StreamResultFuture.createFollower(init.sessionIndex, init.planId, init.streamOperation, init.from, channel, init.pendingRepair, init.previewKind);
-                streamSession = sessionProvider.apply(new SessionIdentifier(init.from, init.planId, init.sessionIndex));
-            }
-            else if (message instanceof IncomingStreamMessage)
-            {
-                // TODO: it'd be great to check if the session actually exists before slurping in the entire stream,
-                // but that's a refactoring for another day
-                StreamMessageHeader header = ((IncomingStreamMessage) message).header;
-                streamSession = sessionProvider.apply(new SessionIdentifier(header.sender, header.planId, header.sessionIndex));
-            }
-
+            StreamSession streamSession = message.getOrCreateSession(channel);
             if (streamSession == null)
                 throw new IllegalStateException(createLogTag(null, channel) + " no session found for message " + message);
```

Comment:
Based on my understanding, calling  `getOrCreateSession` on `StreamInitMessage` or `IncomingStreamMessage` is guaranty to return a `StreamSession` and calling `getOrCreateSession` on another type of message would throw an `UnsupportedOperationException`. By consequence, it seems that this null check is useless. Am I missing something?

Reply:
it shouldn't be null. I will remove it.